### PR TITLE
Roleselection: Remove Role avoidance, Use plymeta:CanSelectRole

### PR DESCRIPTION
**Pull request by EntranceJew**

### CODE REVIEW

**Comment by EntranceJew** [comment]
the last commit gets rid of role avoiding entirely but keeps the mystical bingo ball chance that everyone loves

**Comment by EntranceJew** [comment]
image

it's like Cherie Lily once said: ["icon"](https://youtu.be/wrN7Z9BYKy0?t=125)

File: lua/terrortown/menus/gamemode/gameplay/general.lua [hunk 1]
```diff
@@ -4,6 +4,7 @@ CLGAMEMODESUBMENU.base = "base_gamemodesubmenu"
 
 CLGAMEMODESUBMENU.priority = 100
 CLGAMEMODESUBMENU.title = "submenu_gameplay_general_title"
+CLGAMEMODESUBMENU.icon = Material("vgui/ttt/vskin/helpscreen/gameplay")
```
Comment (ZenBre4ker) [inline lua/terrortown/menus/gamemode/gameplay/general.lua:7]: Please move those changes to another PR

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua [hunk 2]
```diff
@@ -754,7 +754,7 @@ end
 -- @realm server
 -- @deprecated
 function plymeta:GetAvoidDetective()
-	return self:GetAvoidRole(ROLE_DETECTIVE)
+	return false
```
Comment (ZenBre4ker) [inline gamemodes/terrortown/gamemode/server/sv_player_ext.lua:757]: Remove Role Avoidance code completely. Add it as breaking change.
Dont think anybody would use these.

**Review by ZenBre4ker — CHANGES_REQUESTED** [review]
Mostly looks good. Gotta take a deeper look later.

**Review by ZenBre4ker — CHANGES_REQUESTED** [review]
What happened in this Pullrequest? There is a lot useful stuff removed.

File: lua/terrortown/lang/en.lua [hunk 3]
```diff
@@ -2085,6 +2082,9 @@ L.beacon_short_desc = "Beacons are used by policing roles to add local wallhacks
 -- 2023-11-18
 L.entity_pickup_owner_only = "Only the owner can pick this up"
 
+-- 2023-11-20
+-- on this day, nothing happened in this location
+
```
Comment (TimGoll) [inline lua/terrortown/lang/en.lua:2087]: huh, what about this? :D

**Review by TimGoll — APPROVED** [review]
It looks good to me, thanks for taking that job. Ultimately roleselection is @ZenBre4ker's thing so he has the last word on this. But it is good to go from my side

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua (hunk 2 above)
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/server/sv_player_ext.lua:757]: FOr me only the full removal of role avoidance is missing.

File: lua/terrortown/lang/en.lua (hunk 3 above)
Comment (EntranceJew) [reply lua/terrortown/lang/en.lua:2087]: that was mostly so the merge didn't snag on the language files after pulling it from way in the past

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua (hunk 2 above)
Comment (EntranceJew) [reply gamemodes/terrortown/gamemode/server/sv_player_ext.lua]: check again

**Review by saibotk — APPROVED** [review]

**Comment by EntranceJew** [comment]
> It looks good to me, thanks for taking that job. Ultimately roleselection is @ZenBre4ker's thing so he has the last word on this. But it is good to go from my side

should be done now afaik

**Comment by ZenBre4ker** [comment]
Sorry, not home until monday. Looked at it briefly. I noticed a few nitpicks I try to post later, but I will merge it earliest next week after I loaded and tested the code myself. Roleselection is hard to review, but it looks good so far. Thanks :)

File: lua/terrortown/entities/roles/traitor/shared.lua [hunk 4]
```diff
@@ -29,9 +29,11 @@ function ROLE:PreInitialize()
 		pct = 0.4,
 		maximum = 32,
 		minPlayers = 1,
+
 		traitorButton = 1,
+
```
Comment (ZenBre4ker) [inline lua/terrortown/entities/roles/traitor/shared.lua:34]: Remove new lines

File: lua/terrortown/lang/de.lua [hunk 5]
```diff
@@ -975,7 +975,7 @@ L.menu_guide_description = "Hilft dir mit TTT2 zurecht zu kommen und erklärt ei
 L.menu_bindings_description = "Belege Tasten von TTT2 und seinen Erweiterungen nach deinem Geschmack"
 L.menu_language_description = "Stelle die Sprache des Spiels ein"
 L.menu_appearance_description = "Verändere die Erscheinung und Performance der UI"
-L.menu_gameplay_description = "Vermeide Rollen und verändere ein paar andere Funktionen"
+--L.menu_gameplay_description = "Tweak voice and sound volume, accessibility settings, and gameplay settings."
```
Comment (ZenBre4ker) [inline lua/terrortown/lang/de.lua:978]: German translation:
```suggestion
L.menu_gameplay_description = "Verändere Lautstärke-, Barrierefreiheits- und Gameplay-Einstellungen."
```

File: lua/terrortown/lang/de.lua [hunk 6]
```diff
@@ -1002,8 +1002,6 @@ L.submenu_appearance_interface_title = "Interface"
 L.submenu_appearance_miscellaneous_title = "Verschiedenes"
 
 L.submenu_gameplay_general_title = "Allgemein"
-L.submenu_gameplay_avoidroles_title = "Vermeide Rollen"
-
```
Comment (ZenBre4ker) [inline lua/terrortown/lang/de.lua:1006]: Readd newline

File: lua/terrortown/lang/fr.lua [hunk 7]
```diff
@@ -1002,8 +1002,6 @@ L.submenu_appearance_interface_title = "Interface"
 L.submenu_appearance_miscellaneous_title = "Autre"
 
 L.submenu_gameplay_general_title = "Général"
-L.submenu_gameplay_avoidroles_title = "Blacklistage de rôle"
-
```
Comment (ZenBre4ker) [inline lua/terrortown/lang/fr.lua:1006]: newline

File: lua/terrortown/lang/zh_tw.lua [hunk 8]
```diff
@@ -1002,8 +1002,6 @@ L.submenu_appearance_interface_title = "界面"
 L.submenu_appearance_miscellaneous_title = "其他"
 
 L.submenu_gameplay_general_title = "通用"
-L.submenu_gameplay_avoidroles_title = "避免特定身份"
-
```
Comment (ZenBre4ker) [inline lua/terrortown/lang/zh_tw.lua:1006]: newline

**Review by ZenBre4ker — CHANGES_REQUESTED** [review]
Final review hopefully next week.
I hope these are the last changes, I'll require.

Thanks for still doing this, much appreciated.

File: gamemodes/terrortown/gamemode/server/sv_player_ext.lua [hunk 9]
```diff
@@ -765,13 +746,18 @@ end
...
+
+	-- or if the randomness decides
+	if math.random(3) == 2 then return true end
 
-	return (
-		choice_count <= role_count
-		or self:GetBaseKarma() > min_karmas and GAMEMODE.LastRole[self:SteamID64()] == ROLE_INNOCENT
-		or math.random(3) == 2
-	) and (choice_count <= role_count or not self:GetAvoidRole(roleData.index))
+	return false
```
[the remark this answers is not in the document]
Comment (saibotk) [reply gamemodes/terrortown/gamemode/server/sv_player_ext.lua:760]: I would disagree here. This is fine and rather readable.

This way all if statements above are considered the happy exits with the default being a return false.

File: gamemodes/terrortown/gamemode/server/sv_roleselection.lua [hunk 10]
```diff
@@ -826,6 +810,9 @@ function roleselection.SelectRoles(plys, maxPlys)
 
 	plys = roleselection.GetSelectablePlayers(plys or player.GetAll())
 
+	-- Randomize role assignment by shuffling the list early.
+	table.Shuffle(plys)
```
[the remark this answers is not in the document]
Comment (EntranceJew) [reply gamemodes/terrortown/gamemode/server/sv_roleselection.lua:814]: it helps in the sense that players.GetAll() is still in network order, so there is still priority, and it doesn't hurt to have it. if anything does utilize `plys` via a hook between now and the end they will get network priority ordered players, such as the FinalizeRoles hook

**Review by ZenBre4ker — APPROVED** [review]
I guess you are right, I trust you. Lets merge it. Thanks :)
